### ISSUE TITLE
Holes and overlap in vectorization of complex regions for jigsaw puzzle

### ISSUE THREAD

**Issue by proceduraljigsaw** [jankovicsandras/imagetracerjs#59]
I'm developing a jigsaw puzzle generator based on a celular automata. The pieces grow and then the raster image gets vectorized. The jigsaws are meant to be lasercut, so the pieces must tile the whole area with no holes and no overlapping. As far as I understand the library treats each "piece" separately, finds its border and fits splines and lines around it independently for each layer (piece in this case) without taking into account the neighboring layers (pieces). So I think this behavior is inherent to the algorithm. Before trying to create my own usecase-specific border tracing engine, do you think it would be possible to modify the library to fit my usecase without changing it too much? I really love the looks of the pieces when vectorized with the library and I want to get the same look, just without holes and overlapping.

![BaseImage](https://user-images.githubusercontent.com/76650290/131247064-702a0087-9651-4fac-b192-474c71087914.png)
![Vectorized](https://user-images.githubusercontent.com/76650290/131247067-9a08b9f0-ce9f-4323-9abc-9bd8e9013e08.png)

**Comment by jankovicsandras** [comment]
Hi,

This is an interesting question.

One problem is that vectorizing is ambiguous, applying the same logic on same pixel-edges can create either holes or overlappings. Like on this figure: all 4 coloured shapes has the same pixel-edge in the middle, but if we trace them as triangles, there will be overlapping, if we trace them as trapezoids, there will be holes.

![issue 15 figure](https://user-images.githubusercontent.com/12699003/39175559-319488e0-479a-11e8-982c-7a0a8d331fb9.png)

(See also Issue 15)

There might be advanced algorithms that can solve this, but ImageTracer focuses on simplicity and minimalism, so I can't offer a solution now.

A related problem is that when two shapes share a line segment, this segment will be traced twice and it's not guaranteed that tracing results the same curves. I think solving this require a graph of shape-meeting-points, a list of edges between points, and a registry of which shape has which graph edges. Then the edges would be traced and the shapes would be assembled later. As you see, this need another set of algorithms, and is not in the scope of ImageTracer now. The results would also be "spiky" I guess, so while it's possible to guarantee covering the plane, it might look bad.

I have another idea you could try, but I'm not sure it would fit your needs. You could try to render the SVG to canvas (maybe with upscaling), "fix" the overlappings and holes by coloring these pixels to one of the neighbor shapes, then tracing it again. Iterating this process might reduce the errors under the lasercut engineering tolerance.

**Comment by jankovicsandras** [comment]
I hope this helped. Feel free to reopen the issue if you have more questions.

**Comment by proceduraljigsaw** [comment]
Thanks for the explanation. I already figured it out and got it working. First I removed the color quantization as it's not required, I already have an indexed color array. Then I wrote some code to find the "corners" of the pieces and a border tracer that goes from corner to corner following the boundary of the pixels. This yields a set of open paths from corner to corner, each representing each border between pieces. This substitutes the marching squares thing. I feed these paths to the interpolator, which I modified to take into account that it's not working with loops but open paths. Then, straight to the fitter, which is also modified to handle open paths instead of loops. Overall I only use the interpolator and fitter, but it works great. It looks pretty good actually. It doesn't re-build the closed pieces and it doesn't need to, as this is for laser cutting, but I don't think it would be hard to do if needed for aesthetics or as jigsaw solution
![IMG-202108](https://user-images.githubusercontent.com/76650290/131629582-3e7162f7-fad6-4501-926a-dd3760fddbfa.jpg)
. It was very easy to modify and figure out once I started digging in the library, and this is from someone with barely no experience with javascript.
